{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fc114b19-0e16-4bd9-b7c8-4ed0e4b6e012",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3920be13-bfea-450d-aee5-80051989b1ce",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a18160b-03ae-46fb-88a8-6bddf36d3ac1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5db077e1-53fb-4699-864f-840114cb06ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results:\n",
      "{'P07204_TRBM_HUMAN': {'fasta_format': '>sp|P07204|TRBM_HUMAN Thrombomodulin OS=Homo sapiens OX=9606 GN=THBD PE=1 SV=2\\nMLGVLVLGALALAGLGFPAPAEPQPGGSQCVEHDCFALYPGPATFLNASQICDGLRGHLM\\nTVRSSVAADVISLLLNGDGGVGRRRLWIGLQLPPGCGDPKRLGPLRGFQWVTGDNNTSYS\\nRWARLDLNGAPLCGPLCVAVSAAEATVPSEPIWEEQQCEVKADGFLCEFHFPATCRPLAV\\nEPGAAAAAVSITYGTPFAARGADFQALPVGSSAAVAPLGLQLMCTAPPGAVQGHWAREAP\\nGAWDCSVENGGCEHACNAIPGAPRCQCPAGAALQADGRSCTASATQSCNDLCEHFCVPNP\\nDQPGSYSCMCETGYRLAADQHRCEDVDDCILEPSPCPQRCVNTQGGFECHCYPNYDLVDG\\nECVEPVDPCFRANCEYQCQPLNQTSYLCVCAEGFAPIPHEPHRCQMFCNQTACPADCDPN\\nTQASCECPEGYILDDGFICTDIDECENGGFCSGVCHNLPGTFECICGPDSALARHIGTDC\\nDSGKVDGGDSGSGEPPPSPTPGSTLTPPAVGLVHSGLLIGISIASLCLVVALLALLCHLR\\nKKQGAARAKMEYKCAAPSKEVVLQHVRTERTPQRL\\n', 'protein_seq': 'MLGVLVLGALALAGLGFPAPAEPQPGGSQCVEHDCFALYPGPATFLNASQICDGLRGHLMTVRSSVAADVISLLLNGDGGVGRRRLWIGLQLPPGCGDPKRLGPLRGFQWVTGDNNTSYSRWARLDLNGAPLCGPLCVAVSAAEATVPSEPIWEEQQCEVKADGFLCEFHFPATCRPLAVEPGAAAAAVSITYGTPFAARGADFQALPVGSSAAVAPLGLQLMCTAPPGAVQGHWAREAPGAWDCSVENGGCEHACNAIPGAPRCQCPAGAALQADGRSCTASATQSCNDLCEHFCVPNPDQPGSYSCMCETGYRLAADQHRCEDVDDCILEPSPCPQRCVNTQGGFECHCYPNYDLVDGECVEPVDPCFRANCEYQCQPLNQTSYLCVCAEGFAPIPHEPHRCQMFCNQTACPADCDPNTQASCECPEGYILDDGFICTDIDECENGGFCSGVCHNLPGTFECICGPDSALARHIGTDCDSGKVDGGDSGSGEPPPSPTPGSTLTPPAVGLVHSGLLIGISIASLCLVVALLALLCHLRKKQGAARAKMEYKCAAPSKEVVLQHVRTERTPQRL', 'N-glycosylation motif': {'motif': 'N-{P}-[ST]-{P}', 'positions': [47, 115, 116, 382, 409]}}}\n"
     ]
    }
   ],
   "source": [
    "class UniprotRequest:\n",
    "    \n",
    "    \"\"\"\n",
    "    The init takes a uniprot_id_list and will retry 10 times. This is helpful because you are making a call to an external\n",
    "    server and initializing that link for use in the rest of the program. \n",
    "\n",
    "    find_motif\n",
    "\n",
    "    Values passed: \n",
    "    1. Protein sequence\n",
    "    2. REGEX : This link explains how Python3 regex works: https://docs.python.org/3/library/re.html \n",
    "\n",
    "    In the example below, what is the REGEX doing? \n",
    "\n",
    "    Need lookahead in regex to allow overlapping of motifs.\n",
    "\n",
    "    First, it defines the find_motifs() method DIFFERENT than find_motif()\n",
    "    Next there's a fasta_to_seq() method defined\n",
    "    Next there's a git_protein_data() method defined. \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, uniprot_id_list, retry=10):\n",
    "        self.retry = retry\n",
    "        self.uniprot_URL = \"http://www.uniprot.org/uniprot/\"\n",
    "        self.suffix = \".fasta\"\n",
    "        self.results_dict = {}\n",
    "        self.get_protein_data(uniprot_id_list)\n",
    "        \n",
    "    def find_motif(self, protein_sequence, motif_regex):\n",
    "        start_positions = []\n",
    "        matches = re.finditer(motif_regex, protein_sequence)\n",
    "        for match in matches:\n",
    "            start_positions.append(match.start() + 1)  # 1 indexed\n",
    "        return start_positions\n",
    "    \"\"\"\n",
    "    More on REGEX: https://docs.python.org/3/library/re.html\n",
    "    \"\"\"\n",
    "\n",
    "    def find_motifs(self, motif_regex=r\"N{1,1}(?=[^P]{1,1}(S|T){1,1}[^P]{1,1})\", motif=\"N-{P}-[ST]-{P}\",\n",
    "                    motif_name=\"N-glycosylation motif\"):\n",
    "        for item in self.results_dict:\n",
    "            \"\"\"\n",
    "            protein_seq is getting results from the URL and processing them, looking for the\n",
    "            protein_seq part of the data returned. \n",
    "            \"\"\"\n",
    "            protein_seq = self.results_dict[item][\"protein_seq\"]\n",
    "            \"\"\"\n",
    "            Here's the individual call to find_motif for each item in the dictionary\n",
    "            \"\"\"\n",
    "            motif_positions = self.find_motif(protein_seq, motif_regex)\n",
    "            \"\"\"\n",
    "            This returns the results with a decorator (new data) that enumerates where the motif was \n",
    "            found in the protein sequence\n",
    "            This is the position data added by this method in the output below: \n",
    "            47, 115, 116, 382, 409\n",
    "            \n",
    "            you can learn more about this if you select the python flavor here: https://regex101.com/\n",
    "            \"\"\"\n",
    "            self.results_dict[item][motif_name] = {\"motif\": motif, \"positions\": motif_positions}\n",
    "\n",
    "    def fasta_to_seq(self, fasta):\n",
    "        start_index = re.search(r\"\\n\", fasta).start() + 1\n",
    "        sequence = fasta[start_index:]\n",
    "        sequence = re.sub(r\"[\\n\\r]+\", \"\", sequence)\n",
    "        return sequence\n",
    "\n",
    "    def get_protein_data(self, uniprot_id_list):\n",
    "\n",
    "        for item in uniprot_id_list:\n",
    "            data_entry = {}\n",
    "            fasta_format = self.data_transaction(item)\n",
    "            data_entry[\"fasta_format\"] = fasta_format\n",
    "            protein_seq = self.fasta_to_seq(fasta_format)\n",
    "            data_entry[\"protein_seq\"] = protein_seq\n",
    "            self.results_dict[item] = data_entry\n",
    "        return\n",
    "\n",
    "    def data_transaction(self, uniprot_id):\n",
    "        response = requests.get(self.uniprot_URL + uniprot_id + self.suffix)\n",
    "        retry_times = 0\n",
    "        if response.ok:\n",
    "            return response.text\n",
    "        else:\n",
    "            retry_times += 1\n",
    "            if retry_times == self.retry:\n",
    "                response.raise_for_status()\n",
    "\n",
    "    def get_data(self):\n",
    "        return self.results_dict\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    \"\"\"\n",
    "    The section below initializes a variable using UniprotRequest(), then it runs the find_motifs() method \n",
    "    \"\"\"\n",
    "    test = UniprotRequest([\"P07204_TRBM_HUMAN\"])\n",
    "    test.find_motifs()\n",
    "\n",
    "    print(\"results:\")\n",
    "    print(test.results_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6d1f950-c715-45ed-9e58-5b805622a563",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
